### PR TITLE
Support for Factom Identity Keys

### PR DESCRIPTION
**Pull request by samuelvanderwaal**
This PR adds support for the [Factom Identity keys](https://github.com/FactomProject/FactomDocs/blob/master/Identity.md) and fixes one minor typo in `ec.py`'s docstrings.

All tests pass currently but feel free to provide input on any design choices made and reject the PR if you don't like it.

### CODE REVIEW

File: factom_keys/id.py
Comment (sambarnes) [inline factom_keys/id.py]: Rather than generate a list of pairs, I think this function should take the form of: 
```python
def generate_key_pair(level: int):
```
And instantiate the keys with the appropriate level prefix (if its valid). That way you only generate one key pair at the priority level that you want

File: factom_keys/id.py
Comment (sambarnes) [inline factom_keys/id.py]: I would prefer the name be more indicative of the intended use (i.e. that these are specifically for ANO Server Identities).

Something like `ServerIdentityPrivateKey` and `ServerIdentityPublicKey`. That way we do not have any confusion when DIDs are widespread in factom

File: factom_keys/id.py
Comment (sambarnes) [inline factom_keys/id.py]: I feel like this opens up the possibility of user bugs where they instantiate a key with an `sk1...` but then accidentally use the wrong `key_num` here. I think a safer way of doing this would be:
- remove `key_num` parameter here
- instead have a `level: int` that is an attribute on the `ServerIdentityPublicKey` and `ServerIdentityPrivateKey` classes, which this string function will use
- also add a `def set_key_level(self, level: int):` on both classes, checking that its a valid level value

Sure theres still the possibility of a user explicitly changing the levels between a public and private key objects, but that will probably be more rare case

File: factom_keys/id.py
Comment (sambarnes) [inline factom_keys/id.py]: When using a `key_string` rather than bytes, let's have this call the `set_key_level()` function i recommended, using the appropriate level corresponding to key string.

When using the `key_bytes`, we can have it default to level 4 (least powerful key), and they'll just have to call this function manually after creation if they want to use a different level.

File: factom_keys/id.py
Comment (sambarnes) [inline factom_keys/id.py]: Let's have this one first create the `ServerIdentityPublicKey`, then call `set_key_level()` for the public key using the same key level as found on this private key

File: factom_keys/id.py
Comment (sambarnes) [inline factom_keys/id.py]: Same comments I had on the PublicKey object

File: factom_keys/id.py
Comment (sambarnes) [inline factom_keys/id.py]: same comment I had on the Public Key object

**Review by sambarnes — CHANGES_REQUESTED** [review]
Overall looks good, thanks for the contribution! Just some minor things I think could be reorganized for easier usage and less confusion down the line

File: factom_keys/id.py
Comment (samuelvanderwaal) [reply factom_keys/id.py]: Makes sense. This made me realize I need to add some string or regex checking to the `is_valid` methods to ensure we get the prefixes we expect.

File: factom_keys/id.py
Comment (samuelvanderwaal) [reply factom_keys/id.py]: I like `ServerIdentity...` for naming. I'll use that.

File: factom_keys/id.py
Comment (samuelvanderwaal) [reply factom_keys/id.py]: Yeah I agree this makes more sense for the general case. Does it make sense to still have a helper function that generates a complete set of keys? I think it's a common use case for a developer to want to create a complete set of identity keys (sk1, id1) - (sk4, id4).

File: factom_keys/serverid.py
Comment (sambarnes) [inline factom_keys/serverid.py]: In (PEP-8)[https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles] we should be capitalizing ID as an abbreviation, like you had before

> Note: When using acronyms in CapWords, capitalize all the letters of the acronym. Thus HTTPServerError is better than HttpServerError.

File: factom_keys/serverid.py
Comment (sambarnes) [inline factom_keys/serverid.py]: These should inherit from `ValueError` as they are just a more specific form of one. I will change mine elsewhere if I have them like this (and probably just move all exceptions to a shared file, same for the base58 helper functions)

File: factom_keys/serverid.py
Comment (sambarnes) [inline factom_keys/serverid.py]: Rather than baking this into the function, it can just be:

```python
class InvalidParamsError(ValueError):
    pass
```

and usage would be: `raise InvalidParamsError("Only provide one of seed bytes or key_string, not both")` so that we can be flexible with what the error is used for in the future. (I am likely going to move all these errors into a shared errors.py file)

File: factom_keys/serverid.py
Comment (sambarnes) [inline factom_keys/serverid.py]: I think we can remove this safety check here since we already checked that its a valid key a few lines up, so that exception should be impossible to hit. Can just simplify this into: `self.key_level = int(key_string[2])

File: factom_keys/serverid.py
Comment (sambarnes) [inline factom_keys/serverid.py]: Same naming comment

File: tests/test_serverid.py
Comment (sambarnes) [inline tests/test_serverid.py]: Name here should probably reflect the classes, as `TestServerIDKeys`

**Review by sambarnes — CHANGES_REQUESTED** [review]
Just a couple minor things. Some stuff that I saw myself doing and will need change as well after we get this merged in

I really like that `@key_level.setter` decorator you used, I didn't know that was a thing you could do for `@property`s

File: factom_keys/serverid.py
Comment (samuelvanderwaal) [reply factom_keys/serverid.py]: Should I make that shared file now or just change these here and let you merge all the errors to a common file?

Comment (samuelvanderwaal) [reply factom_keys/serverid.py]: Scratch that. I'll let you merge errors into a common file if you wish, and will just make the changes you suggested above.

Comment (sambarnes) [reply factom_keys/serverid.py]: I'll do the shared file myself, just change them here should be fine

**Review by sambarnes — APPROVED** [review]

**Pull request merged**
